File: BOJ(Java 공부용)/5719.java
import java.io.*;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.PriorityQueue;
import java.util.StringTokenizer;
public class Main {
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    static int[] gox = {0, 1, -1, 0, 1, -1};
    static int[] goy = {1, 0, 0, -1, 0, 0};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        while (true) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int n = Integer.parseInt(st.nextToken()), m = Integer.parseInt(st.nextToken());
            if (n == 0 && m == 0) break;
            ArrayList<p>[] graph = new ArrayList[n + 1];
            ArrayList<p>[] back = new ArrayList[n + 1];
            int[] dist = new int[n + 1];
            for (int i = 0; i <= n; i++) {
                graph[i] = new ArrayList<>();
                back[i] = new ArrayList<>();
            }
            st = new StringTokenizer(br.readLine());
            int start = Integer.parseInt(st.nextToken()), end = Integer.parseInt(st.nextToken());
            for (int i = 0; i < m; i++) {
                st = new StringTokenizer(br.readLine());
                int here = Integer.parseInt(st.nextToken()), next = Integer.parseInt(st.nextToken()), cost = Integer.parseInt(st.nextToken());
                graph[here].add(new p(next, cost));
                back[next].add(new p(here, cost));
            }
            dijk(start, end, graph, dist);
            dfs(end, back, graph, dist);
            dijk(start, end, graph, dist);
            if (dist[end] == 1e9) bw.write("-1\n");
            else bw.write(dist[end] + "\n");
        }
        bw.flush();
        bw.close();
    }

    private static void dfs(int here, ArrayList<p>[] back, ArrayList<p>[] graph, int[] dist) {
        for (int i = 0; i < back[here].size(); i++) {
            int b = back[here].get(i).next;
            int backcost = back[here].get(i).cost;
            if (dist[b] + backcost == dist[here]) {
                for (int j = 0; j < graph[b].size(); j++) {
                    int next = graph[b].get(j).next;
                    if (next == here) {
                        graph[b].set(j, new p(here, -1));
                        dfs(b, back, graph, dist);
                    }
                }
            }
        }
    }

    private static void dijk(int start, int end, ArrayList<p>[] graph, int[] dist) {
        PriorityQueue<p> pq = new PriorityQueue<>((a, b) -> {
            if (a.cost < b.cost) return -1;
            else if (a.cost > b.cost) return 1;
            else {
                if (a.next < b.next) return -1;
                else if (a.next > b.next) return 1;
                else return 0;
            }
        });
        pq.add(new p(start, 0));
        Arrays.fill(dist, (int) 1e9);
        dist[start] = 0;
        while (!pq.isEmpty()) {
            int here = pq.peek().next;
            int cost = pq.peek().cost;
            pq.poll();
            if (cost > dist[here]) continue;
            for (int i = 0; i < graph[here].size(); i++) {
                int next = graph[here].get(i).next;
                int nextcost = graph[here].get(i).cost;
                if (nextcost == -1) continue;
                if (dist[next] > dist[here] + nextcost) {
                    dist[next] = dist[here] + nextcost;
                    pq.add(new p(next, dist[next]));
                }
            }
        }
    }

    private static class p {
        int next, cost;

        private p(int next, int cost) {
            this.next = next;
            this.cost = cost;
        }
    }
}

File: BOJ(Python 공부용)/5719.py
import itertools
import sys
from collections import deque
from heapq import heappush, heappop
from math import sqrt

input = sys.stdin


def dijk(st):
    pq = []
    for i in range(n):
        dist[i] = int(1e9)

    heappush(pq, (0, st))
    dist[st] = 0
    while pq:
        cost, here = heappop(pq)
        if dist[here] < cost: continue
        for j in range(len(graph[here])):
            next, nextcost = graph[here][j][0], graph[here][j][1]
            if check[here][j]: continue
            if dist[next] > dist[here] + nextcost:
                dist[next] = dist[here] + nextcost
                heappush(pq, (dist[next], next))


def dfs(en):
    for e in reverseGraph[en]:
        back, backcost = e[0], e[1]
        if dist[back] + backcost == dist[en]:
            for j in range(len(graph[back])):
                next = graph[back][j][0]
                if next == en:
                    check[back][j] = True
                    dfs(back)


if __name__ == '__main__':
    while True:
        n, m = map(int, input.readline().split())
        if n == 0 and m == 0: break
        st, en = map(int, input.readline().split())
        graph = [[] for _ in range(n)]
        check = [[False] * n for _ in range(n)]
        reverseGraph = [[] for _ in range(n)]
        dist = [0] * n
        for _ in range(m):
            here, next, cost = map(int, input.readline().split())
            graph[here].append((next, cost))
            reverseGraph[next].append((here, cost))

        dijk(st)
        dfs(en)
        dijk(st)
        if dist[en] == int(1e9):
            print("-1")
        else:
            print(dist[en])
